sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java

```diff
@@ -117,26 +114,62 @@ public AnalyzedRelation visitMultiSourceSelect(MultiSourceSelect multiSourceSele
 
     @Override
     public AnalyzedRelation visitTwoRelationsUnion(TwoRelationsUnion twoTableUnion, Context context) {
-        process(twoTableUnion.first(), context);
-        process(twoTableUnion.second(), context);
+        QuerySpec querySpec = twoTableUnion.querySpec();
+        QuerySpec firstQuerySpec = twoTableUnion.first().querySpec();
+        QuerySpec secondQuerySpec = twoTableUnion.second().querySpec();
 
         // PushDown limit to the relations of the union since as the union tree is built
         // by the parser the limit exists only in the top level TwoRelationsUnion
-        Optional<Symbol> pushDownLimit = Limits.add(twoTableUnion.querySpec().limit(),
-                                                    twoTableUnion.querySpec().offset());
-
-        assert !twoTableUnion.first().querySpec().limit().isPresent() :
+        Optional<Symbol> pushDownLimit = Limits.add(querySpec.limit(), querySpec.offset());
+        assert !firstQuerySpec.limit().isPresent() :
             "Limit on the individual relations of the union should be empty";
-        twoTableUnion.first().querySpec().limit(pushDownLimit);
-        assert !twoTableUnion.second().querySpec().limit().isPresent() :
+        firstQuerySpec.limit(pushDownLimit);
+        assert !secondQuerySpec.limit().isPresent() :
             "Limit on the individual relations of the union should be empty";
-        twoTableUnion.second().querySpec().limit(pushDownLimit);
+        secondQuerySpec.limit(pushDownLimit);
+
+        // Push down orderBy to the relations of the union for performance optimization
+        replaceFieldReferences(querySpec);
+        Optional<OrderBy> orderBy = querySpec.orderBy();
+        // Convert orderBy symbols referring to the 1st relation of the union to InputColumns
+        // as rootOrderBy must apply to corresponding columns of all relations
+        if (orderBy.isPresent()) {
+            final List<Symbol> outputs = querySpec.outputs();
+            orderBy.get().replace(new com.google.common.base.Function<Symbol, Symbol>() {
+                @Nullable
+                @Override
+                public Symbol apply(@Nullable Symbol symbol) {
+                    return InputColumn.fromSymbol(symbol, outputs);
+                }
+            });
 
-        // TODO: push down order by
-        replaceFieldReferences(twoTableUnion.querySpec());
+            OrderBy newFirstOrderBy = new OrderBy(mergeOrderBy(firstQuerySpec.orderBy(), orderBy));
```

Comment:
I'd suggest adding an new OrderBy.merge(OrderBy orderBy, boolean copy) method to do that